flow/Net2.actor.cpp

```diff
@@ -947,6 +957,7 @@ Net2::Net2(const TLSConfig& tlsConfig, bool useThreadPool, bool useMetrics)
 	  reactor(this),
 	  stopped(false),
 	  tasksIssued(0),
+	  ready(FLOW_KNOBS->READY_QUEUE_RESERVED_SIZE),
```

Comment:
I assume this helps to improve performance? Do we have test number to show how much improvement it brings? @sfc-gh-anoyes

Reply:
We had some empirical evidence that this improves client network thread throughput by ~10%. Our test setup was something like this:

```
#define FDB_API_VERSION 300
#include <foundationdb/fdb_c.h>

#include <atomic>
#include <condition_variable>
#include <cstring>
#include <iostream>
#include <thread>
#include <vector>

void fdb_check(fdb_error_t e) {
  if (e) {
    std::cerr << fdb_get_error(e) << std::endl;
    std::abort();
  }
}

struct DB {
  ~DB() {
    if (db)
      fdb_database_destroy(db);
  }
  FDBDatabase *db = nullptr;
};

struct Tx {
  ~Tx() {
    if (tx)
      fdb_transaction_destroy(tx);
  }
  FDBTransaction *tx = nullptr;
};

struct Future {
  ~Future() {
    if (f)
      fdb_future_destroy(f);
  }
  FDBFuture *f = nullptr;
};

struct Event {
  void wait() {
    std::unique_lock<std::mutex> l(mutex);
    cv.wait(l, [this]() { return this->complete; });
  }
  bool wait_for(double seconds) {
    std::unique_lock<std::mutex> l(mutex);
    cv.wait_for(l, std::chrono::nanoseconds{static_cast<long>(1e9 * seconds)},
                [this]() { return this->complete; });
    return complete;
  }
  void set() {
    std::unique_lock<std::mutex> l(mutex);
    complete = true;
    cv.notify_all();
  }

private:
  std::mutex mutex;
  std::condition_variable cv;
  bool complete = false;
};

FDBDatabase *fdb_open_database(const char *clusterFile) {
  FDBDatabase *db;
  FDBCluster *cluster;
  FDBFuture *f = fdb_create_cluster(clusterFile);
  fdb_check(fdb_future_block_until_ready(f));
  fdb_check(fdb_future_get_cluster(f, &cluster));
  fdb_future_destroy(f);
  f = fdb_cluster_create_database(cluster, (uint8_t *)"DB", 2);
  fdb_check(fdb_future_block_until_ready(f));
  fdb_check(fdb_future_get_database(f, (FDBDatabase **)&db));
  fdb_future_destroy(f);
  fdb_cluster_destroy(cluster);
  return db;
}

int main(int argc, char **argv) {
  fdb_check(fdb_select_api_version(300));
  if (argc != 4) {
    fprintf(stderr,
            "Expected %s <cluster_file> knob_name=value "
            "<path_to_external_client>\n",
            argv[0]);
    return 1;
  }

  // Options
  fdb_check(fdb_network_set_option(
      FDBNetworkOption::FDB_NET_OPTION_TRACE_ENABLE, (const uint8_t *)"", 0));
  fdb_check(fdb_network_set_option(
      FDBNetworkOption::FDB_NET_OPTION_DISABLE_LOCAL_CLIENT,
      (const uint8_t *)"", 0));
  fdb_check(fdb_network_set_option(FDBNetworkOption::FDB_NET_OPTION_KNOB,
                                   (const uint8_t *)argv[2],
                                   std::strlen(argv[2])));
  fdb_check(fdb_network_set_option(
      FDBNetworkOption::FDB_NET_OPTION_EXTERNAL_CLIENT_LIBRARY,
      (const uint8_t *)argv[3], std::strlen(argv[3])));

  fdb_check(fdb_setup_network());
  std::thread network_thread{&fdb_run_network};
  DB db;
  db.db = fdb_open_database(argv[1]);
  {
    Tx tx;
    fdb_check(fdb_database_create_transaction(db.db, &tx.tx));
    for (;;) {
      Future read;
      read.f = fdb_transaction_get_read_version(tx.tx);
      fdb_check(fdb_future_block_until_ready(read.f));
      int64_t v;
      if (!fdb_future_get_version(read.f, &v)) {
        break;
      }
      fdb_transaction_reset(tx.tx);
    }
  }
  std::atomic<int64_t> completions{0};
  std::vector<std::thread> clients(1000);
  Event done;
  std::thread report{[&]() {
    int64_t last_count = 0;
    while (!done.wait_for(1)) {
      int64_t new_count = completions.load();
      printf("%ld ops/s\n", new_count - last_count);
      last_count = new_count;
    }
  }};
  for (auto &client : clients) {
    client = std::thread{[&]() {
      Tx tx;
      fdb_check(fdb_database_create_transaction(db.db, &tx.tx));
      for (int i = 0; i < 1000; ++i) {
        Future read;
        read.f = fdb_transaction_get(tx.tx, (const uint8_t *)"", 0, false);
        Event e;
        fdb_check(fdb_future_set_callback(
            read.f,
            +[](FDBFuture *f, void *p) {
              Event &e = *(Event *)p;
              e.set();
            },
            &e));
        e.wait();
        ++completions;
        fdb_transaction_reset(tx.tx);
      }
    }};
  }
  for (auto &client : clients) {
    client.join();
  }
  fdb_check(fdb_stop_network());
  network_thread.join();
  done.set();
  report.join();
}
```

Which you can compile like this: `c++ saturate_fdb_client.cpp -pthread $BUILDDIR/lib/libfdb_c.so -rpath $BUILDDIR/lib -g -O2` and run like this

```
$ ./a.out
Expected ./a.out <cluster_file> knob_name=value <path_to_external_client>
```

FWIW I tried this just now on master, and I don't see any differences in performance. Another thing we noticed, which I can still reproduce, is that external clients have better throughput than local clients. CC @sfc-gh-kmakino